File: imperative-vs-declarative/src/test/java/ga/overfullstack/imperative/parallel/ForkJoinConcat.java
package ga.overfullstack.imperative.parallel;

import static ga.overfullstack.common.Common.EXPECTED_RESULT;
import static ga.overfullstack.common.Common.TEAM;
import static ga.overfullstack.imperative.ImperativeConcat.concatLastNames;
import static ga.overfullstack.imperative.parallel.Util.AVAILABLE_CORES;
import static ga.overfullstack.imperative.parallel.Util.concatResultsFromForks;
import static org.junit.jupiter.api.Assertions.assertEquals;

import java.util.List;
import java.util.concurrent.ForkJoinPool;
import java.util.concurrent.RecursiveTask;
import org.junit.jupiter.api.Test;

class ForkJoinConcat {

  @Test
  void parallelWithForkJoinPool() {
    var forkJoinPool = new ForkJoinPool(AVAILABLE_CORES);
    var actualResult = forkJoinPool.invoke(new MyRecursiveTask(TEAM));
    System.out.println(actualResult);
    assertEquals(EXPECTED_RESULT, actualResult);
  }

  static class MyRecursiveTask extends RecursiveTask<String> {
    private static final long serialVersionUID = -5978274303314688527L;

    private static final int MIN_TEAM_SIZE = 2; // In real-world, DO NOT have it below 10,000
    private final List<String> team;

    MyRecursiveTask(List<String> team) {
      this.team = team;
    }

    @Override
    protected String compute() {
      if (team.size() > MIN_TEAM_SIZE) {
        var mid = team.size() / 2;
        final var myRecursiveTask1 = new MyRecursiveTask(team.subList(0, mid));
        final var myRecursiveTask2 = new MyRecursiveTask(team.subList(mid, team.size()));
        myRecursiveTask1.fork();
        myRecursiveTask2.fork();

        return concatResultsFromForks(myRecursiveTask1, myRecursiveTask2);
      } else {
        return concatLastNames(team);
      }
    }
  }
}

File: imperative-vs-declarative-kt/src/test/kotlin/ga/overfullstack/imperative/parallel/ForkJoinConcat.kt
/* gakshintala created on 4/14/20 */
package ga.overfullstack.imperative.parallel

import com.google.common.truth.Truth.assertThat
import ga.overfullstack.common.EXPECTED_RESULT
import ga.overfullstack.common.TEAM
import ga.overfullstack.imperative.ImperativeConcat.Companion.concatLastNames
import java.util.concurrent.ForkJoinPool
import java.util.concurrent.RecursiveTask
import org.junit.jupiter.api.Test

internal class MyRecursiveTask(private val team: List<String?>?) : RecursiveTask<String>() {
  override fun compute(): String =
    team?.let {
      if (it.size > MIN_TEAM_SIZE) {
        val mid = it.size / 2
        val myRecursiveTask1 = MyRecursiveTask(it.subList(0, mid))
        val myRecursiveTask2 = MyRecursiveTask(it.subList(mid, it.size))
        myRecursiveTask1.fork()
        myRecursiveTask2.fork()
        concatResultsFromForks(myRecursiveTask1, myRecursiveTask2)
      } else {
        concatLastNames(it) // Using concatLastNames as MIN_TEAM_SIZE can be greater than 2
      }
    } ?: ""

  companion object {
    private const val serialVersionUID = -5978274303314688527L
    private const val MIN_TEAM_SIZE = 2 // In real-world, DO NOT have it below 10,000
  }
}

class ForkJoinConcat {

  @Test
  fun parallelWithForkJoinPool() {
    val forkJoinPool = ForkJoinPool(AVAILABLE_CORES)
    val actualResult = forkJoinPool.invoke(MyRecursiveTask(TEAM))
    println(actualResult)
    assertThat(actualResult).isEqualTo(EXPECTED_RESULT)
  }
}
